1-theory/2-different-paradigms.adoc

```diff
@@ -0,0 +1,217 @@
+
+There are two key types of requiremen for an API:
+
+- RPC (Remote Procedure Call)
+- RDA (Remote Data Access)
+
+RPC allows a client to say "I would like to initiate this procedure/command/function".
+RDA allows a client to say "I would like to access this data".
+
+Pretty much every API in the world does one of these two things. Now, how they go about doing that can vary wildly, and this is where paradigms come in.
+
+== Paradigm
+
+There are a few known paradigms floating around that cover the requirements of "execute a thing" and "interact with data". To oversimplify things a bit, it’s reasonably fair to say that all APIs conform to one of these paradigms:
+
+- RPC
+- REST
+- Query Language
+
+NOTE: Yep, RPC shows up in both lists, because "do a thing" is simple enough it's both a vague concept and a paradigm.
+
+These are general approaches to building APIs, but not a specific tool or specification. They are merely an idea or set of ideas, not a tangible thing. For that, we need to look at implementations.
+
+*Implementations* are something you can actually download, install, and use to build an API, that conforms to whatever rules the implementors picked, from whichever paradigms they wanted to use at the time.
+
+*Specifications* (or standards, recommendations, etc.) are often drafted up by various working groups, to help implementations share functionality in the same way. An API and a client in different languages can work together perfectly if they’re all following the specification correctly. The difference between a standard and a specification is usually just down to who made it. When a working group like the IETF or W3C create something, it's a standard, but when Google knock something out on their own, it's a specification.
+
+For example:
+
+- SOAP is a W3C recommendation, following the RPC paradigm, with implementations like gSOAP
+- gRPC is a implementation, following the RPC paradigm, which has no standard or specification by any working group, but authors Google Inc. did document the protocol
+- REST is a paradigm, which has never been turned into a specification, and has no official implementations, but building a REST API is usually just a case of picking appropriate standards and tooling
+
+Making direct comparisons between any of these things is tough because it should be possible to see that paradigms, implemntations and specifications are all rather different, but let's try and truck on regardless. First, let us try and figure out the main conceptual differences between the paradigms, then we can get stuck into implementations and such later.
+
+== Remote Procedure Call (RPC)
+
+RPC is the earliest, simplest form of API interaction. It is about executing a block of code on another server, and when implemented in HTTP or AMQP it can become a Web API. There is a method and some arguments, and that is pretty much it. Think of it like calling a function in JavaScript, taking a method name and arguments.
```

Comment:
```suggestion
RPC is the earliest, simplest form of API interaction. It is about executing a block of code on another server, and when implemented in HyperText Transfer Protocol (HTTP) or Advanced Message Queuing Protocol (AMQP) it can become a Web API. There is a method and some arguments, and that is pretty much it. Think of it like calling a function in JavaScript, taking a method name and arguments.
```

Reply:
Feel free to dismiss this one, but seeing as this is the first time those acronyms pop up might be worth expanding them and then you're good to use the shortcuts. I knew AMQP was a queue but didn't know what it actually expands to 😳.